rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1089 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+
+### 1.1. GRU
+
+GRU is defined by the following formula:
+
+![GRU-LBR](formula-gru-lbr.png)
+
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+## 2. API Changes
+
+Attention memory descriptor can be built from existing `rnn_desc_t` fields (memory
+descriptors, cell kind, direction, propagation kind, etc.). Taking into account
+this fact and to avoid breaking changes an attention memory descriptor will not
+be part of `dnnl_rnn_desc_t`. Instead a user should use query API to get
+attention shapes.
```

Comment:
I agree that the RNN desc is already bloated and it would be preferable to not add more to it. However, if we want to go with the query route, we have to make sure we will not be introducing a bunch of spurious reorders if we require layouts/type different than what framework use.

If the biggest concern with adding a desc is size, maybe we can put the augru_attention_desc in a union with another incompatible field (e.g. peephole weights)?

Long term, we might have to find a better way than putting everything in the same desc though...

Reply:
Yeah, I would choose alg_kind augru + reusing peephole weights if I have to choose between flags and an algorithm. flags would be very strange way to describe computations. Also it might be the case that we will need to support AIGRU or other modifications of GRU. The only downside with a new algorithm is new api in both C and C++ interfaces.

Reply:
The biggest concern is introducing v2 with all consequences until oneDNN 3.0